### PR TITLE
refactor(common): deprecate `ngStyle` and `ngClass`

### PR DESCRIPTION
**Pull request by JeanMeche**
The overhead of those directives (additional bundle size, distinct import) doesn't outweighs the few additional cases it supports compared to compiler provided `class` and `style` bindings.

DEPRECATE: Both `ngStyle` and `ngClass` directives are deprecated, use the their direct bindings instead.

Also, documentation the migration for this will address #40623. 

---- 
Note: 
The native bindings support almost the same usecases at the exception of 2 :

* They don't support `Set`
* They don't support keys with multiple styles/classes separated with spaces.
* Unit suffix for `style` (eg `myStyles = { 'width.px': 100}`) wouldn't be supported either. 

The recommendation is to drop those space-separated-keys into separate ones and replace Sets with arrays.

### CODE REVIEW

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

**Comment by DmitryEfimenko** [comment]
This is a deprecation of a very widely used feature. If it is to be deprecated, I'd like to see clear examples in the deprecation comments showing how to migrate code where key-value expression was used

**Comment by JeanMeche** [comment]
@DmitryEfimenko key-value expressions are already supported by the existing bindings. 

The numbers are quite important in Google's monorepo, (`[ngClass]` is used twice as much as `[class]`. 

Also a migration will be provided to migrate the cases where is it possible to migrate the templates but it will not be possible to migrate everything as there are some breaking cases that will require manuel updates. 

That being said, this will probably be a soft deprecation.

**Comment by destus90** [comment]
@JeanMeche 

What about [unit suffixes](https://hackmd.io/jzDc7hIDTdWtQblv2TbL9A#Support-for-unit-suffixes-in-ngStyle) in style? I found this link in this PR: #40623

Is it actual or not?

**Comment by dhhyi** [comment]
> The native bindings support almost the same usecases at the exception of 2 :
> ...
> They don't support keys with multiple styles/classes separated with spaces.

This second use case is probably the most common, when using Angular together with styling systems that are heavily based on utility css classes, like tailwind or bootstrap (to some extent). I am a bit worried about the additional bloat that will be necessary to split all used classes and repeat the condition.

Something not really uncommon like
```html
<div [ngClass]="{ 'ring ring-pink-500 ring-offset-2 dark:ring-green-500' : condition }">...
```
Would need to become
```html
<div [class]="{ 'ring' : condition, 'ring-pink-500' : condition, 'ring-offset-2' : condition, 'dark:ring-green-500' : condition }">...
```
Which is really not that great to look at and quite cumbersome to write.

Considering alternatives, you could of course introduce a helper class in the styling of the component and use @apply like this
```css
.my-ring {
  @apply ring ring-pink-500 ring-offset-2 dark:ring-green-500;
}
```
And use the helper class like this
```html
<div [ngClass]="{ 'my-ring' : condition }">...
```
But this would contradict principles of tailwind to have utility and locality first (https://tailwindcss.com/docs/utility-first) and avoid abstractions (https://tailwindcss.com/docs/reusing-styles).

So I am concerned and have a few questions:

Is there another beautiful way of still keeping the awesome experience working with tailwind and Angular I haven't thought about?

Is it possible to enhance Angular to still support keys with multiple styles/classes separated with spaces?

**Comment by e-oz** [comment]
```
<div class=”{{ condition ? 'ring ring-pink-500 ring-offset-2 dark:ring' : '' }}"></div>
```

**Comment by dhhyi** [comment]
> When one of the conditions becomes false, the directive will remove the classes, enlisted for this condition, even if other conditions have these classes too.

@e-oz 
Oh wow, that's an interesting behavior. 

But it's caused by undisciplined programmers, and taming them is a task for eslint.
This seems to be a job for a new rule that could be contributed to [angular-eslint](https://github.com/angular-eslint/angular-eslint).

**Comment by JeanMeche** [comment]
We'll supersed this change with #59240.

TLDR: We acknowledge that `ngClass` and `ngStyle` are still widely used (including inside G3). Not all those usages can be migrated to using the native bindings. 
Instead of a hard deprecation, we'll go toward a soft-deprecation by updating of the recommendations in our docs.
